src/game/common/thing/thingfactory.cpp

```diff
@@ -13,15 +13,167 @@
+    new_template->Set_Is_Allocated();
+    overrid->Set_Next(new_template);
+    return new_template;
+}
+
+void ThingFactory::Free_Database()
+{
+    while (m_firstTemplate) {
+        ThingTemplate *t = m_firstTemplate;
+        m_firstTemplate = t->Friend_Get_Next_Template();
+        t->Delete_Instance();
+    }
+
+    m_templateMap.clear();
+}
+
+void ThingFactory::PostProcessLoad()
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {
+        i->Resolve_Names();
+    }
+}
+
+void ThingFactory::Reset()
+{
+    ThingTemplate *t = m_firstTemplate;
+
+    while (t) {
+        bool first = false;
+        ThingTemplate *t2 = t->Friend_Get_Next_Template();
+
+        if (t == m_firstTemplate) {
+            first = true;
+        }
+
+        Utf8String str(t->Get_Name());
+        Overridable *o = t->Delete_Overrides();
+
+        if (!o) {
+            if (first) {
+                m_firstTemplate = t2;
+            }
+        }
+
+        if (!o) {
+            m_templateMap.erase(str);
+        }
+
+        t = t2;
+    }
+}
+
+void ThingFactory::Add_Template(ThingTemplate *tmplate)
+{
+    if (m_templateMap.find(tmplate->Get_Name()) != m_templateMap.end()) {
+        captainslog_debug("Duplicate Thing Template name found: %s", tmplate->Get_Name());
+    }
+
+    tmplate->Friend_Set_Next_Template(m_firstTemplate);
+    m_firstTemplate = tmplate;
+    m_templateMap[tmplate->Get_Name()] = tmplate;
+}
+
+const ThingTemplate *ThingFactory::Find_Template_By_ID(unsigned short id)
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {
+        if (i->Get_Template_ID() == id) {
+            return i;
+        }
+    }
+
+    captainslog_debug("template %d not found", id);
+    return nullptr;
+}
+
+const ThingTemplate *ThingFactory::Find_Template_Internal(const Utf8String &name, bool b)
+{
+    auto i = m_templateMap.find(name);
+
+    if (i != m_templateMap.end()) {
+        return i->second;
+    }
+
+    if (!strncmp(name, "***TESTING", strlen("***TESTING"))) {
+        ThingTemplate *tmplate = New_Template("Un-namedTemplate");
+        tmplate->Init_For_LTA(name);
+        m_templateMap.erase("Un-namedTemplate");
+        m_templateMap[name] = tmplate;
+        return Find_Template_Internal(name, true);
+    } else {
+        if (b && name.Is_Not_Empty()) {
+            captainslog_debug(
+                "Failed to find thing template %s (case sensitive) This issue has a chance of crashing after you ignore it!",
+                name.Str());
+        }
+        return nullptr;
+    }
+}
+
+Object *ThingFactory::New_Object(const ThingTemplate *tmplate, Team *team, BitFlags<OBJECT_STATUS_COUNT> status_bits)
 {
 #ifdef GAME_DLL
-    return Call_Method<ThingTemplate *, ThingFactory, Utf8String>(PICK_ADDRESS(0x004B0B50, 0x00407DD0), this, name);
+    return Call_Method<Object *, ThingFactory, const ThingTemplate *, Team *, BitFlags<OBJECT_STATUS_COUNT>>(

```

Comment:
Are there any details you can put into a TODO as to why the function can't be done at this time.

Reply:
Fixed.